src/Runner.Worker/ExpressionFunctions/HashFiles.cs

```diff
@@ -0,0 +1,139 @@
+using GitHub.DistributedTask.Pipelines.ObjectTemplating;
+using GitHub.Runner.Common;
+using GitHub.Runner.Sdk;
+using System.Reflection;
+using System.Threading;
+using System.Runtime.Serialization;
+
+namespace GitHub.Runner.Worker.Handlers
+{
+    public class FunctionTrace : ITraceWriter
+    {
+        private GitHub.DistributedTask.Expressions2.ITraceWriter _trace;
+
+        public FunctionTrace(GitHub.DistributedTask.Expressions2.ITraceWriter trace)
+        {
+            _trace = trace;
+        }
+        public void Info(string message)
+        {
+            _trace.Info(message);
+        }
+
+        public void Verbose(string message)
+        {
+            _trace.Info(message);
+        }
+    }
+
+    public sealed class HashFiles : Function
+    {
+        protected sealed override Object EvaluateCore(
+            EvaluationContext context,
+            out ResultMemory resultMemory)
+        {
+            resultMemory = null;
+
+            // hashFiles() only works on the runner and only works with files under GITHUB_WORKSPACE
+            // Since GITHUB_WORKSPACE is set by runner, I am using that as the fact of this code runs on server or runner.
+            if (context.State is DistributedTask.ObjectTemplating.TemplateContext templateContext &&
+                templateContext.ExpressionValues.TryGetValue(PipelineTemplateConstants.GitHub, out var githubContextData) &&
+                githubContextData is DictionaryContextData githubContext &&
+                githubContext.TryGetValue(PipelineTemplateConstants.Workspace, out var workspace) == true &&
+                workspace is StringContextData workspaceData)
+            {
+                string searchRoot = workspaceData.Value;
+                bool followSymlink = false;
+                string pattern = "";
+                if (Parameters.Count == 1)
+                {
+                    pattern = Parameters[0].Evaluate(context).ConvertToString();
+                }
+                else
+                {
+                    if (string.Equals(Parameters[0].Evaluate(context).ConvertToString(), "--followSymbolicLinks", StringComparison.OrdinalIgnoreCase))
+                    {
+                        followSymlink = true;
+                    }
+
+                    pattern = Parameters[1].Evaluate(context).ConvertToString();
+                }
+
+
+                // Convert slashes on Windows
+                if (s_isWindows)
+                {
+                    pattern = pattern.Replace('\\', '/');
+                }
+
+                context.Trace.Info($"Search root directory: '{searchRoot}'");
+                context.Trace.Info($"Search pattern: '{pattern}'");
+
+                string binDir = Path.GetDirectoryName(Assembly.GetEntryAssembly().Location);
+                string runnerRoot = new DirectoryInfo(binDir).Parent.FullName;
+
+                string node = Path.Combine(runnerRoot, "externals", "node12", "bin", "node");
```

Comment:
consider `$"node{IOUtil.ExeExtension}"` then you can remove `s_isWindows`

Reply:
will fix.